test/opt/assembly_builder.h

```diff
@@ -0,0 +1,207 @@
+                                      vec_asm_code.begin(), vec_asm_code.end());
+    return *this;
+  }
+
+  // Get the SPIR-V assembly code in string.
+  std::string GetCode() const {
+    std::stringstream ss;
+    for (auto& line : global_preamble_vec_) {
+      ss << line << std::endl;
+    }
+    for (auto& line : names_vec_) {
+      ss << line << std::endl;
+    }
+    for (auto& line : types_consts_globals_vec_) {
+      ss << line << std::endl;
+    }
+    for (auto& line : main_func_preamble_vec_) {
+      ss << line << std::endl;
+    }
+    for (auto& line : main_func_entry_block_vec_) {
+      ss << line << std::endl;
+    }
+    for (auto& line : main_func_postamble_vec_) {
+      ss << line << std::endl;
+    }
+    return ss.str();
+  }
+
+ private:
+  // Adds a given name to the Name section with OpName. If the given name has
+  // beed added before, do nothing.
+  void AddOpNameIfNotExist(const std::string& name) {
+    if (!names_.count(name)) {
+      std::stringstream opname_inst;
+      opname_inst << "OpName "
+           << "%" << name << " \"" << name << "\"";
+      size_t pos = names_vec_.size();
+      names_vec_.emplace_back(opname_inst.str().c_str());
+      names_[name] = pos;
+    }
+  }
+
+  // Adds the names in a vector of assembly code strings to the Names section.
+  // If a '=' sign is found in an instruction, this instruction will be treated
+  // as an ID defining instruction. The ID name used in the instruction will be
+  // extracted and added to the Names section.
+  void AddNamesFor(const std::vector<std::string>& vec_asm_code) {
+    for (const auto& line : vec_asm_code) {
+      if (line.find('=') != std::string::npos) {
+        size_t assign_sign = line.find('=');
+        std::string name = line.substr(0, assign_sign);
+        name.erase(remove_if(name.begin(), name.end(),
+                             [](char c) { return c == ' ' || c == '%'; }),
+                   name.end());
+        AddOpNameIfNotExist(name);
+      }
+    }
+  }
+
+  // The vector that contains common preambles shared across all test SPIR-V
+  // code.
+  std::vector<std::string> global_preamble_vec_;
+  // The vector that contains the code that declare names (aka. the Names
+  // section).
+  std::vector<std::string> names_vec_;
+  // The vector that contains the code to declare types, constants and global
+  // variables (aka. the Types-Constants-Globals section).
+  std::vector<std::string> types_consts_globals_vec_;
+  // The vector that contains the preamble of main function.
+  std::vector<std::string> main_func_preamble_vec_;
+  // The vector that contains the code in main function's entry block.
+  std::vector<std::string> main_func_entry_block_vec_;
+  // The vector that contains the postamble of main function body.
+  std::vector<std::string> main_func_postamble_vec_;
+  // All of the defined variable names.
+  std::unordered_map<std::string, size_t> names_;
```

Comment:
It maps to the position of the corresponding OpName instruction in the name_ (original names_vec_) vector. Currently just using it as a set, the position is not used, but I was thinking about removing instructions (now I think removing might not be necessary at all).